### PR TITLE
Add finalizer to User PlacementRule

### PR DESCRIPTION
**Pull request by BenamarMk**
This PR solves issue #288 . It adds a finalizer to the user PlacementRule and watches for PlacementRule deletion.
If the User PlacementRule for a specific instance of DRPC is deleted, then the DRPC cleanup process is executed

### CODE REVIEW

File: controllers/drplacementcontrol_controller_test.go [hunk 1]
```diff
@@ -893,6 +899,17 @@ var _ = Describe("DRPlacementControl Reconciler", func() {
 				Expect(userPlacementRule.Status.Decisions[0].ClusterName).To(Equal(EastManagedCluster))
 			})
 		})
+
+		When("Deleting user PlacementRule", func() {
+			It("Should cleanup DRPC", func() {
+				// ----------------------------- DELETE DRPC from PRIMARY --------------------------------------
+				By("\n\n*** DELETE User PlacementRule ***\n\n")
+				deleteUserPlacementRule()
```
Comment (BenamarMk) [inline controllers/drplacementcontrol_controller_test.go:907]: For some reason, deleting the User PlacementRule from this unit test does not trigger a notification.  Still investigating.

**Review by raghavendrabhat — APPROVED** [review]
LGTM

**Review by raghavendrabhat — APPROVED** [review]

File: controllers/drplacementcontrol_controller.go [hunk 2]
```diff
@@ -519,11 +588,18 @@ func (r *DRPlacementControlReconciler) processDeletion(ctx context.Context,
 			return ctrl.Result{}, err
 		}
 
-		// Remove DRPCFinalizer. The object will be deleted once
-		// the finalizer is removed
+		// Remove DRPCFinalizer from User PlacementRule.
```
Comment (ShyamsundarR) [inline controllers/drplacementcontrol_controller.go:591]: This block to remove userPl finalizer should be outside the if block?

I see 2 reasons for this,
1. Order of finalizer addition is userPl followed by DRPC, so removal should be the reverse
2. If userPl finalizer is added, but DRPC finalizer addition errored out and userPl was deleted, we will never remove the finalizer on the userPl

**Review by ShyamsundarR — COMMENTED** [review]
LGTM except finalizer removal order comment if that seems correct.

File: controllers/drplacementcontrol_controller.go (hunk 2 above)
Comment (BenamarMk) [reply controllers/drplacementcontrol_controller.go:591]: good catch.  So the finalizer addition of the uPlRule should be after the finalizer addition of the DRPC.  Then in the removal, we will remove the uPlRule first

Comment (BenamarMk) [reply controllers/drplacementcontrol_controller.go:594]: done

Comment (BenamarMk) [reply controllers/drplacementcontrol_controller.go:594]: So let me think about this as I am writing. 
DRPC is deleted, we remove the usrPlRule finalizer, then let's say we fail to update the DRPC object, so we reconcile again.  If we assume that the usrPlRule is deleted by the time we enter the Reconcile function, then we are stuck in that reconciliation loop failing to get usrPlRule.  True, there is a window of trouble.

Comment (BenamarMk) [reply controllers/drplacementcontrol_controller.go:594]: hmmm, but the reverse is also not good. We remove the DRPC finalizer first, then we fail to remove the usrPlRule.  Doesn't that also leak the usrPlRule?

Comment (BenamarMk) [reply controllers/drplacementcontrol_controller.go:595]: So I pushed a new commit (in order to show the changed differences) to solve the above issue.

**Review by ShyamsundarR — APPROVED** [review]

**Pull request merged**